activerecord/lib/active_record/associations/association_scope.rb

```diff
@@ -153,14 +153,17 @@ def add_constraints(scope, owner, association_klass, refl, chain_head, chain_tai
               scope.order_values |= item.order_values
             end
 
-            reflection = reflection.next
+            reflection = next_reflection
           end
 
           scope
         end
 
-        def eval_scope(klass, scope, owner)
-          klass.unscoped.instance_exec(owner, &scope)
+        def eval_scope(klass, table, scope, owner)
+          klass.unscoped.tap { |relation|
+            predicate_builder = PredicateBuilder.new(TableMetadata.new(klass, table))
+            relation.instance_variable_set(:@predicate_builder, predicate_builder)
```

Comment:
We have two choices.

* Expose a setter in the relation.
* Use `Relation.create` instead of `klass.unscoped` like `JoinAssociation`. https://github.com/rails/rails/blob/505463c859927408665ce141426f098fea303a4e/activerecord/lib/active_record/associations/join_dependency/join_association.rb#L48-L49

Which do you think is better?